src/Microsoft.Extensions.WebEncoders.Core/UrlPathDecoder.cs

```diff
@@ -0,0 +1,326 @@
+
+                if ((v & 0xC0) != 0x80)
+                {
+                    return 0;
+                }
+
+                // append the lower 6bit
+                currentDecodeBits = (currentDecodeBits << 6) | (v & 0x3F);
+
+                bytesBuffer[bytesCount - remainingBytes] = (byte)v;
+                remainingBytes--;
+
+                if (remainingBytes == 1 && currentDecodeBits >= 0x360 && currentDecodeBits <= 0x37F)
+                {
+                    // This is going to end up in the range of 0xD800-0xDFFF UTF-16 surrogates that are not allowed in UTF-8;
+                    return 0;
+                }
+                if (remainingBytes == 2 && currentDecodeBits >= 0x110)
+                {
+                    // This is going to be out of the upper Unicode bound 0x10FFFF.
+                    return 0;
+                }
+            }
+
+            if (currentDecodeBits < expectValueMin)
+            {
+                // Overlong encoding (e.g. using 2 bytes to encode something that only needed 1).
+                return 0;
+            }
+
+            try
+            {
+                // Convert the code point into char
+                count = _utf8.Value.GetCharCount(bytesBuffer, 0, bytesCount);
+                if (count > output.Length)
+                {
+                    output = new char[count];
+                }
+
+                _utf8.Value.GetChars(bytesBuffer, 0, bytesCount, output, 0);
+                return bytesCount * 3;
+            }
+            catch (DecoderFallbackException)
+            {
+                return 0;
+            }
+        }
+
+        /// <summary>
+        /// Copy characters in source string to target char array. After copied the indices points at the beginning
+        /// offset of both source and destination are updated to point to the position after the last copied character
+        /// </summary>
+        private static int Copy(string source, ref int sourceStart, char[] destination, ref int destinatonStart, int length)
+        {
+            for (var i = 0; i < length; ++i)
+            {
+                destination[destinatonStart++] = source[sourceStart++];
+
+                if (destinatonStart >= destination.Length ||
+                    sourceStart >= source.Length)
+                {
+                    return i + 1;
+                }
+            }
+
+            return length;
+        }
+
+        /// <summary>
+        /// Find the next % in the sequence. If % is not found, return the sequence length.
+        /// </summary>
+        private static int GetNextEncoded(int start, string source)
+        {
+            var next = source.IndexOf('%', start);
+            if (next == -1)
+            {
+                return source.Length;
+            }
+            else
+            {
+                return next;
+            }
+        }
+
+        private static bool TryGetUnescapedAscii(char first, char second, out char result)
```

Comment:
This name and purpose are out of sync. You don't fail if it's > 0x7F and you keep doing bitwise operation on the resulting char. How about changing it to TryGetUnescapedByte with an out byte?